File: memoria.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <ctime>

using namespace std;

void mostrarTabuleiro(const vector<string>& estado) {
    for (const auto& s : estado) {
        cout << s << " ";
    }
    cout << endl;
}

int main() {
    vector<string> tabuleiro = {"A", "A", "B", "B", "C", "C", "D", "D"};
    srand(time(0));
    random_shuffle(tabuleiro.begin(), tabuleiro.end());

    vector<string> estado(8, "_");
    int acertos = 0;

    while (acertos < tabuleiro.size() / 2) {
        mostrarTabuleiro(estado);
        int pos1, pos2;
        cout << "Escolha a primeira posição (0 a 7): ";
        cin >> pos1;
        cout << "Escolha a segunda posição (0 a 7): ";
        cin >> pos2;

        if (pos1 != pos2 && estado[pos1] == "_" && estado[pos2] == "_") {
            estado[pos1] = tabuleiro[pos1];
            estado[pos2] = tabuleiro[pos2];
            mostrarTabuleiro(estado);

            if (tabuleiro[pos1] == tabuleiro[pos2]) {
                cout << "Par encontrado!" << endl;
                acertos++;
            } else {
                estado[pos1] = "_";
                estado[pos2] = "_";
                cout << "Não combinam!" << endl;
            }
        } else {
            cout << "Escolha posições válidas e diferentes." << endl;
        }
    }
    cout << "Parabéns! Jogo concluído." << endl;
    return 0;
}

File: memoria.py
import random

# Tabuleiro do jogo com pares de letras
tabuleiro = ["A", "A", "B", "B", "C", "C", "D", "D"]
random.shuffle(tabuleiro)
estado = ["_"] * len(tabuleiro)

# Função para exibir o tabuleiro
def mostrar_tabuleiro():
    print(" ".join(estado))

# Jogo da memória
def jogo_memoria():
    acertos = 0
    while acertos < len(tabuleiro) // 2:
        mostrar_tabuleiro()
        pos1 = int(input("Escolha a primeira posição (0 a 7): "))
        pos2 = int(input("Escolha a segunda posição (0 a 7): "))
        if pos1 != pos2 and estado[pos1] == "_" and estado[pos2] == "_":
            estado[pos1], estado[pos2] = tabuleiro[pos1], tabuleiro[pos2]
            mostrar_tabuleiro()
            if tabuleiro[pos1] == tabuleiro[pos2]:
                print("Par encontrado!")
                acertos += 1
            else:
                estado[pos1], estado[pos2] = "_", "_"
                print("Não combinam!")
        else:
            print("Escolha posições válidas e diferentes.")

jogo_memoria()
